<!DOCTYPE html>
<!--[if IE 8]><html class="no-js lt-ie9" lang="en" > <![endif]-->
<!--[if gt IE 8]><!--> <html class="no-js" lang="en" > <!--<![endif]-->
<head>
  <meta charset="utf-8">
  <meta http-equiv="X-UA-Compatible" content="IE=edge">
  <meta name="viewport" content="width=device-width, initial-scale=1.0">
  
  
  <link rel="shortcut icon" href="../img/favicon.ico">
  <title>User guide - Gene Set Analysis Matlab</title>
  <link href='https://fonts.googleapis.com/css?family=Lato:400,700|Roboto+Slab:400,700|Inconsolata:400,700' rel='stylesheet' type='text/css'>

  <link rel="stylesheet" href="../css/theme.css" type="text/css" />
  <link rel="stylesheet" href="../css/theme_extra.css" type="text/css" />
  <link rel="stylesheet" href="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/styles/github.min.css">
  
  <script>
    // Current page data
    var mkdocs_page_name = "User guide";
    var mkdocs_page_input_path = "userguide.md";
    var mkdocs_page_url = null;
  </script>
  
  <script src="../js/jquery-2.1.1.min.js" defer></script>
  <script src="../js/modernizr-2.8.3.min.js" defer></script>
  <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/highlight.min.js"></script>
  <script src="//cdnjs.cloudflare.com/ajax/libs/highlight.js/9.12.0/languages/matlab.min.js"></script>
  <script>hljs.initHighlightingOnLoad();</script> 
  
</head>

<body class="wy-body-for-nav" role="document">

  <div class="wy-grid-for-nav">

    
    <nav data-toggle="wy-nav-shift" class="wy-nav-side stickynav">
      <div class="wy-side-nav-search">
        <a href=".." class="icon icon-home"> Gene Set Analysis Matlab</a>
        <div role="search">
  <form id ="rtd-search-form" class="wy-form" action="../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" title="Type search term here" />
  </form>
</div>
      </div>

      <div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="main navigation">
	<ul class="current">
	  
          
            <li class="toctree-l1">
		
    <a class="" href="..">Overview</a>
	    </li>
          
            <li class="toctree-l1 current">
		
    <a class="current" href="./">User guide</a>
    <ul class="subnav">
            
    <li class="toctree-l2"><a href="#gsam-user-guide">GSAM User Guide</a></li>
    
        <ul>
        
            <li><a class="toctree-l3" href="#example-1-single-gsa-liver-hepatocellular-carcinoma">Example 1: Single GSA, Liver hepatocellular carcinoma</a></li>
        
            <li><a class="toctree-l3" href="#example-2-using-gene-set-collections-derived-from-gems">Example 2: Using gene set collections derived from GEMs</a></li>
        
        </ul>
    

    </ul>
	    </li>
          
        </ul>
      </div>
      &nbsp;
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap">

      
      <nav class="wy-nav-top" role="navigation" aria-label="top navigation">
        <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
        <a href="..">Gene Set Analysis Matlab</a>
      </nav>

      
      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="breadcrumbs navigation">
  <ul class="wy-breadcrumbs">
    <li><a href="..">Docs</a> &raquo;</li>
    
      
    
    <li>User guide</li>
    <li class="wy-breadcrumbs-aside">
      
    </li>
  </ul>
  <hr/>
</div>
          <div role="main">
            <div class="section">
              
                <h1 id="gsam-user-guide">GSAM User Guide</h1>
<p>The following examples demonstrate the use of the GSAM package for common analysis scenarios.</p>
<p>The data files used for these examples can be retrieved from the <code>data/</code> directory on the associated <a href="https://github.com/JonathanRob/GeneSetAnalysisMatlab-doc/tree/master/data">documentation repository</a>. If you wish you run these examples yourself, first download the files in that directory to your local machine and ensure that they are on your current MATLAB path.</p>
<h2 id="example-1-single-gsa-liver-hepatocellular-carcinoma">Example 1: Single GSA, Liver hepatocellular carcinoma</h2>
<h3 id="load-the-data-and-gene-set-collection-file">Load the data and gene set collection file</h3>
<p>Load the differential expression (DE) results for hepatocellular carcinoma (LIHC) (paired-normal tissue vs. primary tumor tissue).</p>
<pre><code class="matlab">DEdata = readtable('DEresults_LIHC.txt');
</code></pre>

<p>Take a look at the table contents.</p>
<pre><code class="matlab">head(DEdata)

% ans =
% 
%   8×4 table
% 
%          gene_IDs           gene_names       logFC         PValue  
%     ___________________    ____________    __________    __________
% 
%     {'ENSG00000000003'}    {'TSPAN6'  }      -0.26345      0.050906
%     {'ENSG00000000419'}    {'DPM1'    }    -0.0030857       0.97084
%     {'ENSG00000000457'}    {'SCYL3'   }      0.044663       0.63024
%     {'ENSG00000000460'}    {'C1orf112'}        0.7297    0.00021919
%     {'ENSG00000000938'}    {'FGR'     }       -1.2409    3.6644e-15
%     {'ENSG00000000971'}    {'CFH'     }      -0.72639    0.00069828
%     {'ENSG00000001036'}    {'FUCA2'   }      0.050749       0.66557
%     {'ENSG00000001084'}    {'GCLC'    }      -0.60349    1.1461e-06
</code></pre>

<p>Extract the data into individual variables</p>
<pre><code class="matlab">geneIDs   = DEdata.gene_IDs;     % gene Ensembl IDs
geneNames = DEdata.gene_names;   % gene abbreviations
log2FC    = DEdata.logFC;        % log2(fold-changes) (normal vs. tumor)
pvals     = DEdata.PValue;       % fold-change significance (p-values)
</code></pre>

<p>Load the <code>KEGG_GSC.gmt</code> gene set collection (GSC) file using the <code>importGSC</code> function, and look at the first few lines to see how it's formatted.</p>
<pre><code class="matlab">gsc = importGSC('KEGG_GSC.gmt');

% Gene set collection contains 146 gene sets and 4966 unique genes.

gsc(1:5,:)

% ans =
%
%   5×2 cell array
%
%     {'KEGG-GLYCOLYSIS-GLUCONEOGENESIS'}    {'ACSS2'}
%     {'KEGG-GLYCOLYSIS-GLUCONEOGENESIS'}    {'GCK'  }
%     {'KEGG-GLYCOLYSIS-GLUCONEOGENESIS'}    {'PGK2' }
%     {'KEGG-GLYCOLYSIS-GLUCONEOGENESIS'}    {'PGK1' }
%     {'KEGG-GLYCOLYSIS-GLUCONEOGENESIS'}    {'PDHB' }
</code></pre>

<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The <code>KEGG_custom_GSC.gmt</code> file used in this example is a modified version of the KEGG gene set collection retrieved from <a href="http://software.broadinstitute.org/gsea/msigdb/genesets.jsp?collection=CP:KEGG">MSigDB</a>, where all disease-related gene sets were removed.</p>
</div>
<h3 id="run-the-gene-set-analysis">Run the gene set analysis</h3>
<p>Run the gene set analysis using the <code>geneSetAnalysis</code> function.</p>
<pre><code class="matlab">GSAres = geneSetAnalysis(geneNames, pvals, log2FC, gsc, 'Wilcoxon', 50000, [20, Inf]);

% Gene set collection contains 146 gene sets and 4966 unique genes.
% Checking for empty gene sets... Removed 0 empty sets.
% Checking for duplicated rows in GSC... Removed 0 duplicated rows.
% Checking gene set sizes... Removed 27 gene sets not satisfying size limits.
% Final number of gene sets remaining: 119
% Subsetting data for mixed-directional calculations... Done.
% Calculating test statistic... Done.
% Calculating significance via gene shuffling... Done.
</code></pre>

<p>Here, we chose the Wilcoxon rank-sum test as the method for combining gene-level statistics, set the number of permutations for significance calculation to 50000, and excluded gene sets that contained less than 20 genes.</p>
<div class="admonition important">
<p class="admonition-title">Important</p>
<p>We used <code>geneNames</code> (instead of e.g. <code>geneIDs</code>) as input to the <code>geneSetAnalysis</code> function because gene names are used in the <code>gsc</code> that was loaded. It is important that the gene names or IDs in the gene list are of the same type as those used in the GSC file.</p>
</div>
<h3 id="view-andor-export-the-results">View and/or export the results</h3>
<p>The results are returned as the <code>GSAres</code> table, which contains information about the size (number of genes) and significance (p-values, raw and adjusted) for each of the evaluated gene sets. The table can be viewed by double-clicking on <code>GSAres</code> in the Workspace, or using the <code>open</code> function.</p>
<pre><code class="matlab">open('GSAres');
</code></pre>

<table>
<thead>
<tr>
<th>GS_name &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp;</th>
<th>GS_size</th>
<th>stat_nondir</th>
<th>p_nondir</th>
<th>padj_nondir</th>
<th>stat_mixup</th>
<th>p_mixup</th>
<th>padj_mixup</th>
<th>stat_mixdn</th>
<th>p_mixdn</th>
<th>padj_mixdn</th>
<th>stat_distup</th>
<th>p_distup</th>
<th>padj_distup</th>
<th>stat_distdn</th>
<th>p_distdn</th>
<th>padj_distdn</th>
</tr>
</thead>
<tbody>
<tr>
<td>KEGG-GLYCOLYSIS-GLUCONEOGENESIS</td>
<td>51</td>
<td>519848</td>
<td>0.0102</td>
<td>0.0371</td>
<td>62259</td>
<td>0.5766</td>
<td>0.8788</td>
<td>196844</td>
<td>0.0018</td>
<td>0.0092</td>
<td>302878</td>
<td>0.9998</td>
<td>1</td>
<td>571415</td>
<td>0.00018</td>
<td>0.00067</td>
</tr>
<tr>
<td>KEGG-CITRATE-CYCLE-TCA-CYCLE</td>
<td>30</td>
<td>294717</td>
<td>0.0832</td>
<td>0.1799</td>
<td>44790</td>
<td>0.3987</td>
<td>0.8041</td>
<td>101736</td>
<td>0.0759</td>
<td>0.1966</td>
<td>200384</td>
<td>0.9817</td>
<td>1</td>
<td>313906</td>
<td>0.0183</td>
<td>0.0496</td>
</tr>
<tr>
<td>KEGG-PENTOSE-PHOSPHATE-PATHWAY</td>
<td>21</td>
<td>198224</td>
<td>0.2115</td>
<td>0.3872</td>
<td>47870</td>
<td>0.2579</td>
<td>0.6929</td>
<td>51476</td>
<td>0.3019</td>
<td>0.5239</td>
<td>176522</td>
<td>0.5622</td>
<td>1</td>
<td>183481</td>
<td>0.4379</td>
<td>0.6596</td>
</tr>
<tr>
<td>KEGG-PENTOSE-AND-GLUCURONATE-INTERCONVERSIONS</td>
<td>21</td>
<td>189170</td>
<td>0.3415</td>
<td>0.5080</td>
<td>8507</td>
<td>0.8360</td>
<td>0.9871</td>
<td>84792</td>
<td>0.2356</td>
<td>0.4344</td>
<td>103850</td>
<td>0.9998</td>
<td>1</td>
<td>256153</td>
<td>0.0002</td>
<td>0.0007</td>
</tr>
<tr>
<td>KEGG-FRUCTOSE-AND-MANNOSE-METABOLISM</td>
<td>32</td>
<td>278495</td>
<td>0.4365</td>
<td>0.6076</td>
<td>74220</td>
<td>0.2803</td>
<td>0.6948</td>
<td>65609</td>
<td>0.6202</td>
<td>0.8317</td>
<td>283091</td>
<td>0.3800</td>
<td>1</td>
<td>265485</td>
<td>0.6201</td>
<td>0.8580</td>
</tr>
<tr>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
<td>...</td>
</tr>
</tbody>
</table>
<p>The tabular format of the GSA results can often be difficult to interpret. A more intuitive way to visualize the results is to use a heatmap. Generate a heatmap to visualize <code>GSAres</code> using the <code>GSAheatmap</code> function.</p>
<pre><code class="matlab">GSAheatmap(GSAres, true, 'pval', 0.01);
</code></pre>

<p><img alt="LIHC GSA heatmap" src="../img/GSAheatmap_LIHC.png" /></p>
<p>Here, we chose to use the adjusted gene-set p-values (as specified with the second input <code>true</code>), and filtered the gene sets to only show those with an adjusted p-value ≤ 0.01 for at least one of the five directionality types.</p>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>For more information on the meaning and interpretation of the different p-value classes (distinct directional down and up, mixed-directinal down and up, and non-directional), see the <a href="https://www.ncbi.nlm.nih.gov/pubmed/23444143">publication</a> where they were introduced. </p>
</div>
<p>Alternatively, the results can be written to a file.</p>
<pre><code class="matlab">writetable(GSAres,'GSAresults.txt','Delimiter','\t');  % export to tab-delimited .txt file
writetable(GSAres,'GSAresults.csv','Delimiter',',');  % export to comma-delimited .csv file
</code></pre>

<h2 id="example-2-using-gene-set-collections-derived-from-gems">Example 2: Using gene set collections derived from GEMs</h2>
<p>Genome-scale metabolic models (GEMs) are convenient collections of all known reactions comprising the metabolism of an organism or cell type. Since most GEMs contain gene information, we can extract gene set collections from the associations of these genes with other model components, such as metabolites or subsystems (pathways).</p>
<h3 id="load-the-gem-and-extract-some-gscs">Load the GEM and extract some GSCs</h3>
<p>For this example, we will use a human genome-scale metabolic model, Human-GEM. A <code>.mat</code> version of the model can be retrieved from the <a href="https://github.com/SysBioChalmers/Human-GEM/tree/master/ModelFiles/mat">Human-GEM GitHub repository</a>.</p>
<p>Once the <code>HumanGEM.mat</code> file is downloaded and added to the current working folder (or somewhere else on the MATLAB path), load it into MATLAB.</p>
<pre><code class="matlab">load('HumanGEM.mat');  % the model is loaded as a structure named 'ihuman'
</code></pre>

<p>Extract a metabolite-based GSC from the model using the <code>extractMetaboliteGSC</code> function.</p>
<pre><code class="matlab">gsc_met = extractMetaboliteGSC(ihuman);

% Gene set collection contains 3715 gene sets and 3625 unique genes.
</code></pre>

<p>Just as other GSCs (such as the KEGG collection described above) associate biological functions or processes to sets of genes, the metabolite GSC associates metabolites to sets of genes. This provides a convenient link between gene-level information and metabolism.</p>
<p>Run a GSA with the same LIHC DE data as in Example 1 above, but now using the metabolite GSC <code>gsc_met</code>. We will also change the gene set size limits to exclude metabolites associated with more than 200 genes, and use the <a href="https://www.pnas.org/content/102/8/2685">"Reporter" method</a> for combining gene-level statistics.</p>
<pre><code class="matlab">GSAres = geneSetAnalysis(geneIDs, pvals, log2FC, gsc_met, 'Reporter', 50000, [20, 200]);

% Checking for empty gene sets... Removed 71 empty sets.
% Checking for duplicated rows in GSC... Removed 0 duplicated rows.
% Checking gene set sizes... Removed 3236 gene sets not satisfying size limits.
% Final number of gene sets remaining: 408
% Subsetting data for mixed-directional calculations... Done.
% Calculating test statistic... Done.
% Calculating significance via gene shuffling... Done.
</code></pre>
              
            </div>
          </div>
          <footer>
  
    <div class="rst-footer-buttons" role="navigation" aria-label="footer navigation">
      
      
        <a href=".." class="btn btn-neutral" title="Overview"><span class="icon icon-circle-arrow-left"></span> Previous</a>
      
    </div>
  

  <hr/>

  <div role="contentinfo">
    <!-- Copyright etc -->
    
  </div>

  Built with <a href="http://www.mkdocs.org">MkDocs</a> using a <a href="https://github.com/snide/sphinx_rtd_theme">theme</a> provided by <a href="https://readthedocs.org">Read the Docs</a>.
</footer>
      
        </div>
      </div>

    </section>

  </div>

  <div class="rst-versions" role="note" style="cursor: pointer">
    <span class="rst-current-version" data-toggle="rst-current-version">
      
      
        <span><a href=".." style="color: #fcfcfc;">&laquo; Previous</a></span>
      
      
    </span>
</div>
    <script>var base_url = '..';</script>
    <script src="../js/theme.js" defer></script>
      <script src="../search/main.js" defer></script>

</body>
</html>
